Fix off by small amounts in sent amount kaspa

coderofstuff opened the pull request:
**Problem:** Floating point math causes inconsistencies when converting kas to sompi.

**Solution:** Use a method that parses the amount as a string, the converts it to sompi string then parse to uint64

**Motivation:** Someone (some 3rd party integrating with kaspa) sent 33184.1489732 via kaspawallet and the other party got 33184.14897319 because the kaspawallet send command loses some precision when converting kas to sompi (floating point math kinda stuff), which is off by 0.00000001. Small difference but irreconcilable to accounting.

tmrlvi commented on cmd/kaspawallet/send.go:
```diff
@@ -99,3 +99,15 @@ func send(conf *sendConfig) error {
+
+	parts := strings.Split(amountInStr, ".")
+
+	convertedAmount, _ := strconv.ParseUint(strings.Join(parts, ""), 10, 64)
```
Check the error value (and return error is it is set). Number could be too big, and then ParseUint will return (nil, error)

coderofstuff replied:
Addressed in [83c3d44](https://github.com/kaspanet/kaspad/pull/2220/commits/83c3d4458eb886ef2272130026d35ece56cbf946)

tmrlvi commented:
`conf.SendAmount` is accepted as float64, so another source of math inconsistencies would be in the conversions of string -> float64 -> string again. 

Maybe we should consider accepting string value, validating, and working directly with it (fixed precision type would be better, but I am not familiar with such solutions in go)

tmrlvi reviewed (changes requested):
`conf.SendAmount` is accepted as float64, so another source of math inconsistencies would be in the conversions of string -> float64 -> string again.

Maybe we should consider accepting string value, validating, and working directly with it (fixed precision type would be better, but I am not familiar with such solutions in go)

tmrlvi commented on cmd/kaspawallet/utils/validate_amount.go:
```diff
@@ -0,0 +1,34 @@
+	}
+
+	// If it parses properly, then this is valid
+	_, err = strconv.ParseFloat(amount, 64)
```
Isn't this redundant (assuming fixed regex)? (though it doesn't hurt)

coderofstuff replied:
It's redundant now that the regex is fixed

tmrlvi commented on cmd/kaspawallet/utils/validate_amount.go:
```diff
@@ -0,0 +1,34 @@
+ */
+func ValidateAmountFormat(amount string) error {
+	// Check whether it's an integer, or a float with max 8 digits
+	match, err := regexp.MatchString("^\\d{1,19}(.\\d{0,8})?$", amount)
```
Also, should it be only 11 digits before the period? (in total 19)

coderofstuff replied:
The maximum possible sompi for uint64 should be `18446744073709551615` - in float that'd be `184467440737.09551615` (12 digits to the left of the decimal). So the rule `"^\([1-9]d{0,11}|0)(\.\d{0,8})?$"` is enough.

More test cases added.

tmrlvi commented on cmd/kaspawallet/send.go:
```diff
@@ -35,7 +35,13 @@ func send(conf *sendConfig) error {
+			return err
+		}
+
+		sendAmountSompi = parsedAmountSompi
```
Is there a reason it is not directly `sendAmountSompi, err := utils.KasToSompi(conf.SendAmount)`?

coderofstuff replied:
Yeah, because of shadowing: https://go.dev/play/p/vAX1AHDpmvs

However, we since `err` is already declared above the lines, I can do (and have done):

```
sendAmountSompi, err = utils.KasToSompi(conf.SendAmount)
```

(note that it's `=` and not `:=`)

tmrlvi commented on cmd/kaspawallet/config.go:
```diff
@@ -296,17 +296,17 @@ func parseCommandLine() (subCommand string, config interface{}) {
-	if (!conf.IsSendAll && conf.SendAmount == 0) ||
-		(conf.IsSendAll && conf.SendAmount > 0) {
+	if (!conf.IsSendAll && conf.SendAmount == "") ||
+		(conf.IsSendAll && conf.SendAmount > "") {
```
You probably want to change this as well to `!=` (to be consistent with below, but both work)

tmrlvi commented:
Seems good. I think you should fix the regex to raise parsing error earlier with a clearer description.

The other two comments are not material, just for cleaner code

tmrlvi reviewed (commented):
Just change

coderofstuff committed:
Consistent config handling

coderofstuff committed:
Deal with SendAmount as strings all the way

coderofstuff committed:
Set variables directly from utils.KasToSompi

Use = instead of := to ensure no shadowing

coderofstuff committed:
Fix off by small amounts in sent amount kaspa

Floating point math causes inconsistencies when converting kas to sompi.

Use a method that parses the amount as a string, the converts it to
sompi then parse back to uint64

coderofstuff committed:
Fix validate amount regex

someone235 commented on cmd/kaspawallet/utils/format_kas.go:
```diff
@@ -14,3 +16,28 @@ func FormatKas(amount uint64) string {
+	amountStr := ""
+
+	if len(parts) == 2 {
+		amountStr = fmt.Sprintf("%s%-*s", parts[0], 8, parts[1]) // Padded with spaces at the end to fill for missing decimals: Sample "0.01234    "
```
Use `math.Log10(SompiPerKaspa)` to determine amount of zeroes (and also add a test that checks `SompiPerKaspa` is a multiple of 10.

someone235 commented on cmd/kaspawallet/utils/format_kas.go:
```diff
@@ -14,3 +16,28 @@ func FormatKas(amount uint64) string {
+		amountStr = fmt.Sprintf("%s%-*s", parts[0], 8, parts[1]) // Padded with spaces at the end to fill for missing decimals: Sample "0.01234    "
+		amountStr = strings.ReplaceAll(amountStr, " ", "0")      // Make the spaces be 0s. Sample "0.012340000"
+	} else {
+		amountStr = fmt.Sprintf("%s00000000", parts[0])
```
Use `math.Log10(SompiPerKaspa)` to determine amount of zeroes

someone235 commented on cmd/kaspawallet/utils/validate_amount.go:
```diff
@@ -0,0 +1,26 @@
+ * 1. May be an integer (no decimal components)
+ * 2. May be float with up to 8 decimal places
+ */
+func ValidateAmountFormat(amount string) error {
```
I wouldn't make it public, and rename it to reflect that it talks about KAS amount (maybe `validateKASAmountFormat`)

someone235 replied:
I would also move it to `format_kas.go` since it's only relevant for `KasToSompi`

someone235 commented on cmd/kaspawallet/utils/format_kas.go:
```diff
@@ -30,14 +33,36 @@ func KasToSompi(amount string) (uint64, error) {
+	match, err := regexp.MatchString("^([1-9]\\d{0,11}|0)(\\.\\d{0,8})?$", amount)
+
+	if !match {
+		return errors.Errorf("Invalid send amount")
```
Change to "invalid amount" (we're not in wallet context)

someone235 commented on cmd/kaspawallet/utils/format_kas.go:
```diff
@@ -30,14 +33,36 @@ func KasToSompi(amount string) (uint64, error) {
 	return convertedAmount, err
 }
+
+func validateAmountFormat(amount string) error {
```
change to `validateKASAmountFormat` (this function is accessible from the whole package so its name should be understandable in any context)

coderofstuff committed:
Use decimal places as defined by constants

Also check if SompiPerKaspa is multiple of 10

coderofstuff committed:
Minor updates for context clarity
